docs/developer-experience/going-further.md

```diff
@@ -0,0 +1,184 @@
+
+One way to achieve this is to leverage environment variables, with untracked `.env` file part of the solution being injected in the Dev Container.
+
+Consider the following files structure:
+
+```bash
+My Application  # main repo directory
+└───.devcontainer
+|       ├───Dockerfile
+|       ├───docker-compose.yml
+|       ├───devcontainer.json
+└───config
+|       ├───.env
+|       ├───.env-sample
+```
+
+The file `config/.env-sample` is a tracked file where anyone can find environment variables to set (with no values, obviously):
+
+```bash
+TENANT_ID=
+SUBSCRIPTION_ID=
+```
+
+Then, each developer who clones the repository can create the file `config/.env` and fills it in with the appropriate values.
+
+Under the folder `.devcontainer`, we introduce a `docker-compose.yml` file to be able to inject this `.env` file
+
+```yaml
+version: '3'
+services:
+  my-workspace:
+    env_file: ../config/.env
+    build:
+      context: .
+      dockerfile: Dockerfile
+    command: sleep infinity
+```
+
+And in order to use the `docker-compose.yml` file, we need to adjust `devcontainer.json`:
+
+```json
+{
+    "name": "My Application",
+    "dockerComposeFile": ["docker-compose.yml"],
+    "service": "my-workspace"
+    ...
+}
+```
+
+This approach can be applied for many other tools by preparing what would be required. The idea is to simplify developers' life and new developers joining the project.
+
+## Custom tools
+
+While working on a project, any developer might end up writing a script to automate a task. This script can be in `bash`, `python` or whatever scripting language they are comfortable with.
+
+Let's say you want to ensure that all markdown files written are validated against specific rules you have set up. As we have seen above, you can include in your Dev Container the tool [markdownlint](https://github.com/DavidAnson/markdownlint). Having the tool installed does not mean developer will know how to use it!
+
+Consider the following solution structure:
+
+```bash
+My Application  # main repo directory
+└───.devcontainer
+|       ├───Dockerfile
+|       ├───docker-compose.yml
+|       ├───devcontainer.json
+└───scripts
+|       ├───check-markdown.sh
+└───.markdownlint.json
+```
+
+The file `.devcontainer/Dockerfile` installs [markdownlint](https://github.com/DavidAnson/markdownlint)
+
+```dockerfile
+...
+RUN apt-get update \
+    && export DEBIAN_FRONTEND=noninteractive \
+    && apt-get install -y nodejs npm
+
+# Add NodeJS tools
+RUN npm install -g markdownlint-cli
```

Comment:
The same could also be achieved using a `postCreateCommand` script.  It would be interesting to see when you would opt to update the docker file vs calling a script after the container has been created.

Reply:
That is a good question. Having what you can install in the Dockerfile will be cached if you need to rebuild - whereas I guess a script will always execute regardless of the state.

Our assumption was: whatever you can install with package managers (apt, npm, pip etc ...) let's do it in the Dockerfile. Eveyrthing that you cannot / or that required a check for the host architecture (Azure-cli, Terraform for instance), let's do it in a bash post command.

Also, we decided not to use `postCreateCommand` to execute our script but call it from the Dockerfile because we wanted to remain as agnostic as possible as not everybody was using vscode ... With that approach we can still share a common dev environment, letting developers choose their IDE

Reply:
How do you think I should proceed?